Make `AtomicRef::new` `const fn`

yvt opened the pull request:
Turns `AtomicRef::new` into `const fn`, making it callable in constant contexts. ~~This currently requires the use of a nightly-only feature, so it's feature-gated by the Cargo feature `nightly`. This can be removed in the future when this unstable feature is stabilized.~~

~~Raises the minimum supported Rust version to 1.32.0 (regardless of the `nightly` feature).~~
As of 5e5a05703892e0f039500203ca7d1dcc26ed4098, this PR raises the minimum supported Rust version to 1.46.0.

This PR depends on #1.

yvt committed:
Replace `AtomicUsize` with `AtomicPtr<T>`

Raises the minimum supported Rust version to 1.32.0.

yvt committed:
Make `AtomicRef::new` `const fn`

Requires the following unstable features:

 - `const_fn` for constructing `PhantomData<&mut _>`
   <https://github.com/rust-lang/rust/issues/57563>
 - `const_if_match` for using `match` in a user-defined `const fn`
   <https://github.com/rust-lang/rust/issues/49146>

yvt committed:
Implement an alternative solution for the variance issue not involving `Mutex`

yvt committed:
Get rid of dependency on `std`

yvt committed:
Make the use of unstable features non-mandatory

yvt committed:
Unnecessitate `#![feature(const_fn)]`

mystor commented:
This feature is on-route to being stabilized in the near future (https://github.com/rust-lang/rust/pull/72437). It might be nice to avoid the nightly feature here by waiting for that feature to be stabilized first, and releasing a `0.2.0` which requires that rustc version, and gets rid of unnecessary APIs like `static_atomic_ref!`.

If you need const constructors right now, it might be easier to add:

```rust
const fn null() -> Self {
    AtomicRef {
        data: AtomicPtr::new(ptr::null_mut()),
        _marker: PhantomData,
    }
}

const fn from(r: &'a T) -> Self {
    AtomicRef {
        data: AtomicPtr::new(r as *const T as *mut T),
        _marker: PhantomData,
    }
}
```

yvt committed:
Remove `nightly` Cargo feature

All required unstable features have been stable since Rust version
1.46.0.

yvt committed:
Remove `ATOMIC_U8_REF_INIT` and `static_atomic_ref`

yvt committed:
Merge remote-tracking branch 'upstream/master' into feat-const-fn

yvt commented:
I added commits that got rid of `nightly` feature since `const_if_match` was stabilized in 1.46.0.

mystor commented on src/lib.rs:
```diff
@@ -73,117 +71,42 @@
 /// Re-export `core` for `static_atomic_ref!` (which may be used in a
 /// non-`no_std` crate, where `core` is unavailable).
 #[doc(hidden)]
 pub use core::{mem as core_mem, ops as core_ops};
```
These re-exports are no longer necessary, as we no longer export a macro.

yvt committed:
Remove unused reexports

These reexports are no longer necessary since we removed the
`static_atomic_ref` macro in 7b6f17e3.

mystor reviewed (approved):
Thanks!
